### PR TITLE
ENH: reorganization of functions for pysat_sgp4

### PR DESCRIPTION
**Pull request by jklenzing**
# Description

Addresses #1.

Reorganizes the functions within this package so that they are reusable for other instruments.
- Adds methods files for each empirical model used, as well as spacecraft coordinate transformations.
- Moves plot method to separate method file.
- Adds some basic tests for the empirical model method files.
- Adds `pysat_ephem` instrument for the Ephem-based propagations
- Removes python 2.7 testing since required pyglow does not support this.
- Adds .codeclimate.yml file to help analyze simplification of functions.
- BUG: apexpy metadata and data name mismatch

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality
      to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?

pytest -vs

Needs plot comparisons between develop and here to validate similar output.

**Test Configuration**:
* Mac OS X 10.14.6
* Python 3.7.3

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``master``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

**Comment by jklenzing** [comment]
To do:
- [ ] add comparison plots to old version
- [ ] merge in #10 
- [ ] add description of adding pyglow object to satellite data
- [ ] rewrite basemap part of plot

**Comment by jklenzing** [comment]
OK, my plan is to leave this pull request alone right now.  We probably need to discuss at the next tag up.  Download it, give it a spin, tell me what you think of the new breakdown.

Note that the plot routines still need to be updated to use cartopy, not basemap.  Given the number of changes to this branch, I think that should be a separate pull request.

Unit tests are not thorough, but shake out most of the code to make sure things run.  Travis-ci is running into server issues on the set up.  These are not consistent over multiple documentation changes, just luck of the draw.

**Comment by rstoneback** [comment]
I'm currently stuck. I can't seem to import pysatMissionPlanning,

import pysatMissionPlanning.methods.aacgmv2

ImportErrorTraceback (most recent call last)
<ipython-input-7-a51cbe11e8bc> in <module>()
----> 1 import pysatMissionPlanning.methods.aacgmv2

/Users/russellstoneback/git/pysatMissionPlanning/pysatMissionPlanning/__init__.py in <module>()
      3 
      4 import os
----> 5 from pysatMissionPlanning import instruments, methods
      6 
      7 __all__ = ['instruments', 'methods']
/Users/russellstoneback/git/pysatMissionPlanning/pysatMissionPlanning/instruments/__init__.py in <module>()
      4 """
      5 
----> 6 from pysatMissionPlanning.instruments import pysat_ephem, pysat_sgp4
      7 
      8 __all__ = ['pysat_ephem', 'pysat_sgp4']
/Users/russellstoneback/git/pysatMissionPlanning/pysatMissionPlanning/instruments/pysat_ephem.py in <module>()
     16 import pandas as pds
     17 import pysat
---> 18 import pysatMissionPlanning.methods.simulators as sim
     19 
     20 # pysat required parameters
/Users/russellstoneback/git/pysatMissionPlanning/pysatMissionPlanning/methods/__init__.py in <module>()
      4 """
      5 
----> 6 from pysatMissionPlanning.methods import aacgmv2
      7 from pysatMissionPlanning.methods import apexpy
      8 from pysatMissionPlanning.methods import plot
ImportError: cannot import name aacgmv2

**Comment by jklenzing** [comment]
Hmm.  Not getting this on my end.  Are you using python 2.7?

**Comment by JonathonMSmith** [comment]
Not sure I've set this up properly, but I installed pysat from pip before installing pysatMissionPlanning and everything runs fine.
I haven't really poked around yet, however.
I tried just installing pysatMissionPlanning without pysat and was getting some weird errors in the install from numpy

**Comment by jklenzing** [comment]
Tried fresh install here on both python 3.6 and 3.7, both working fine.  2.7 fails at the pyglow install.

**Comment by rstoneback** [comment]
Thanks for checking Jeff. I am using 2.7. I’ll be putting together an updated development environment as soon as my new computer comes in.

On Jan 6, 2020, at 2:11 PM, Jeff Klenzing <notifications@github.com> wrote:

﻿

Tried fresh install here on both python 3.6 and 3.7, both working fine. 2.7 fails at the pyglow install.

**Comment by JonathonMSmith** [comment]
So i can get everything up and running easily enough on python 3.7. However, I would like to note a few things that had to be done for that to work.
I had to manually install:
-numpy
-apexpy
-pyglow
-future

All of this is fine, but I wanted to make a note of it.
Another thing about pyglow, it's probably not our responsibility to inform people of this, but if you blindly use pip install pyglow, it will install a package, however it's this: https://pypi.org/project/PyGlow/
Which is obviously not what we need here.

**Comment by jklenzing** [comment]
TODO:
- [ ] Update install instructions for pyglow separate install

**Comment by jklenzing** [comment]
Having looked at this for a while, I think we need an adjustment to the organization.  The methods section should just have bits that interact with other codes, or in the case of `spacecraft`, velocity and unit vector calculations.  The two that don't fit here are `plot` and `simulators`.  `plot` should go up one level.  `simulators` is basically the equivalent of the old cdaweb_methods files.  Not sure pysat_methods is the right name for this.

Proposed structure
-------------------
- pysatMissions
  - instruments
    - pysat_ephem
    - pysat_methods (formerly simulators)
    - pysat_sgp4
  - methods
    - aacgmv2
    - apexpy
    - pyglow
    - spacecraft
  - plot

Thoughts?

**Comment by jklenzing** [comment]
Also, given the difficulties with the pyglow install, should we have this as a "soft" import?

**Comment by rstoneback** [comment]
Perhaps supporting_methods is a reasonable name for simulators. I agree it shouldn't be under methods.

Soft import of pyglow worked well for me. I was able to import pysatMissions (yay!) without pyglow installed at all.

What do you think about adding a pysatMagVect part to methods? Provided by importing the pysatMagVect.satellite submodule.

File: docs/introduction.rst [hunk 1]
```diff
@@ -2,12 +2,13 @@
...
 
-pysatMissionPlanning allows users to run build simulated satellites for TLE info and add empirical data. It includes the pysat_sgp4 instrument module which can be imported into pysat.
+pysatMissionPlanning allows users to run build simulated satellites for TLE info and add empirical data. It includes the pysat_sgp4 and pysat_ephem instrument modules which can be imported into pysat.
 
 Main Features
 -------------
 - Simulate satellite orbits from TLEs and add data from empirical models
 - Import ionosphere and thermosphere values through pyglow
 - Import coordinates through apexpy
+- Import coordinates through aacgmv2
```
Comment (rstoneback) [inline docs/introduction.rst]: - Import geomagnetic basis vectors through pysatMagVect

**Comment by rstoneback** [comment]
Currently trying to check outputs of code with pyglow installed. I'm getting a bunch of errors but I think that is from pyglow. Still working on it.

**Comment by rstoneback** [comment]
pyglow continues to give me problems but those problems have nothing to do with this pull. Looks good. Thanks Jeff.

**Comment by jklenzing** [comment]
> Perhaps supporting_methods is a reasonable name for simulators. I agree it shouldn't be under methods.
> 
I like this better.  Another option is to hide this under a core function, since these are not directly called by the user.  Thinking of something like `_core` with `_list_files` and `_download` as hidden methods to map to `list_files` and `download`.

> Soft import of pyglow worked well for me. I was able to import pysatMissions (yay!) without pyglow installed at all.

Hooray! 🎉 
> 
> What do you think about adding a pysatMagVect part to methods? Provided by importing the pysatMagVect.satellite submodule.

I like this.  Will look into the structure.

File: docs/introduction.rst (hunk 1 above)
Comment (jklenzing) [reply docs/introduction.rst]: Good call!

**Comment by jklenzing** [comment]
Going to merge this in once tests are done.  Thanks for the comments.  Will add a new issue for pysatMagVect.

**Pull request merged**